libc/docs/search.rst

```diff
@@ -0,0 +1,85 @@
+=============
+Search Tables
+=============
+
+.. include:: check.rst
+
+---------------
+Source location
+---------------
+
+-   The main source for bitwise utility functions is located at:
+    ``libc/src/search``.
+
+-   Hashtable implementation is located at:
+    ``libc/src/__support/HashTable``.
+
+-   The tests are located at:
+    ``libc/test/src/search/``.
+
+---------------------
+Implementation Status
+---------------------
+
+POSIX Standard Types
+====================
+
+============================ =========
+Type Name                    Available
+============================ =========
+ACTION                       |check|
+ENTRY                        |check|
+VISIT                        
+============================ =========
+
+
+GNU Extension Types
+===================
+
+============================ ================= =========
+Type Name                    Associated Macro  Available
+============================ ================= ========= 
+struct qelem                 
+__compar_fn_t                __COMPAR_FN_T
+comparison_fn_t
+__action_fn_t                __ACTION_FN_T
+__free_fn_t
```

Comment:
Forgive my naivete, but which functions make use of these types?

Reply:
`qelem` is for `insque` and `remque`. It is the suggested struct header. The remaining is for the binary tree operations. GNU just gives them a name. Notice that `__compar_fn_t` is the same for `qsort`, that's why it has a macro guard.

Reply:
Thanks! So for `qelem`, my read of the man page is that that is historical minutia; I suspect POSIX made these functions take `void*` specifically to avoid this.  Mentioning them at all might lead users or fellow implementers to believe that these are on the TODO list.  To be clear, they are not.

As such I think we should either:
- Mark that we will not implement these.
- Simply do not mention them at all.

---

> The remaining are for the binary tree operations.

Those are not mentioned in any man pages AFAICT. Are you getting these identifiers from glibc sources?

---

> Notice that __compar_fn_t is the same as the one for qsort

Also not mentioned in any man pages. Where are you seeing any mention of `__compar_fn_t`?

Reply:
They were from glibc's headers. 

I put them on just in case users use them. However, I think that ppl writing portable code should not use them.

Reply:
How about mentioning that "the following types provided by GNU libc's header will not be declared by LLVM libc"

Reply:
> However, I think that ppl writing portable code should not actually use them.

100%.  Let's not encourage them to look up what these even are then, lest they come to depend on them.

> How about mentioning that "the following types provided by GNU libc's header will not be declared by LLVM libc"

I don't think it will scale to mention every type internal to every other libc (or even just glibc).

---

> As such I think we should either:
> Mark that we will not implement these.
> Simply do not mention them at all.

Now that I have a better sense of where these originate from, I'm going to change my preferences to just a lone preference for "do not mention them at all."

It was worth discussing, so I am happy to think more about this.

Reply:
I agree to remove "history burden" in a whole 👍🏻 .